--- Day 1: Trebuchet?! ---

Something is wrong with global snow production, and you've been selected to take a look. The Elves have even given you a map; on it, they've used stars to mark the top fifty locations that are likely to be having problems.

You've been doing this long enough to know that to restore snow operations, you need to check all fifty stars by December 25th.

Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

You try to ask why they can't just use a weather machine ("not powerful enough") and where they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot of questions") and hang on did you just say the sky ("of course, where do you think snow comes from") when you realize that the Elves are already loading you into a trebuchet ("please hold still, we need to strap you in").

As they're making the final adjustments, they discover that their calibration document (your puzzle input) has been amended by a very young Elf who was apparently just excited to show off her art skills. Consequently, the Elves are having trouble reading the values on the document.

The newly-improved calibration document consists of lines of text; each line originally contained a specific calibration value that the Elves now need to recover. On each line, the calibration value can be found by combining the first digit and the last digit (in that order) to form a single two-digit number.

For example:

```
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
```

In this example, the calibration values of these four lines are 12, 38, 15, and 77. Adding these together produces 142.

Consider your entire calibration document. What is the sum of all of the calibration values?

In [1]:
# import the regular expression library
import re

# Read the input file from data/aoc_2023_data_01.txt
with open("../data/aoc_2023_data_01.txt") as f:
    data = f.read().splitlines()

# for each entry in data, remove the [A-Z] and [a-z] characters
# and convert the remaining string to an integer

data2 = [re.sub('[A-Za-z]', '', s)for s in data]

# When the length of the string is greater than 1, then read only first and last
# character. Otherwise, for length 1, read only repeat the character twice.
# Convert to Integer
data3 = [int(s[0]+s[-1]) if len(s)>1 else int(s*2) for s in data2]

sum(data3)

54390

## **Your puzzle answer was: 54390**

--- Part Two ---
Your calculation isn't quite right. It looks like some of the digits are actually spelled out with letters: one, two, three, four, five, six, seven, eight, and nine also count as valid "digits".

Equipped with this new information, you now need to find the real first and last digit on each line. For example:

two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
In this example, the calibration values are 29, 83, 13, 24, 42, 14, and 76. Adding these together produces 281.

What is the sum of all of the calibration values?

Your puzzle answer was 54277.

Both parts of this puzzle are complete! They provide two gold stars: **

In [2]:
# For the data list, convert the strings "zero", "one", "two", ..., "nine"
# to their corresponding digit characters.

digits_hash = {"one": "1", "two": "2", "three": "3", "four": "4",
                "five": "5", "six": "6", "seven": "7", "eight": "8", "nine": "9"}

# For each entry in data, replace the digit words with their corresponding digit
# characters. Use lookahead to make sure that digit numbers with shared characters
# are recognized

# The matchobj.group() has to have the 1 argument, otherwise it will return the
# entire match object, because want it to return only the matched string.

data2b = [re.sub("(?=("+'|'.join(digits_hash.keys())+"))", lambda m: digits_hash[m.group(1)], s) for s in data]
data2c = [re.sub('[A-Za-z]', '', s)for s in data2b]
data2d = [int(s[0]+s[-1]) if len(s)>1 else int(s*2) for s in data2c]

print(sum(data2d))

54277


In [3]:
data_example = ["two1nine", "eightwothree","abcone2threexyz","xtwone3four","4nineeightseven2","zoneight234","7pqrstsixteen"]
print(sum(
    [int(s[0]+s[-1]) if len(s)>1 else int(s*2) for s in 
        [re.sub('[A-Za-z]', '', s) for s in 
            [re.sub("(?=("+'|'.join(digits_hash.keys())+"))", lambda m: digits_hash[m.group(1)], s) for s in data_example]
            ]]))


[re.sub("(?=("+'|'.join(digits_hash.keys())+"))", lambda m: digits_hash[m.group(1)], s) for s in data_example]

281


['2two19nine',
 '8eigh2two3three',
 'abc1one23threexyz',
 'x2tw1one34four',
 '49nine8eight7seven2',
 'z1on8eight234',
 '7pqrst6sixteen']

## **Answer: 54277**

## **That's the right answer! You are one gold star closer to restoring snow operations.**

### You have completed Day 1! 